### PR TITLE
[toup] zephyr: crypto: Add PRF-256 Wrapper based on PSA APIs

### PR DESCRIPTION
**Pull request by VivekUppunda**
This commit adds PRF-256 wrapper API for section 11.6.1.7.2 in 802.11ac using PSA APIs. The wrapper API has been validated via test vectors derived during connection and has provided the desired output.

### CODE REVIEW

**Comment by VivekUppunda** [comment]
@fengming-ye - Hi, can you kindly review this Wrapper API? This is only for PRF-256. It is not generic as hmac_prf_bits.

**Comment by fengming-ye** [comment]
Code LGTM. But why don't we add PSA wrapper to hmac_prf_bits?

**Comment by VivekUppunda** [comment]
> Code LGTM. But why don't we add PSA wrapper to hmac_prf_bits?

Presently, I am unable to get test vectors for SHA-384 and SHA-512. If that were available, we could extend this API wrapper to support the same or write wrappers for SHA-384/SHA-512 similar to this API. 

Thanks and regards
Vivek

**Comment by fengming-ye** [comment]
> > Code LGTM. But why don't we add PSA wrapper to hmac_prf_bits?
> 
> Presently, I am unable to get test vectors for SHA-384 and SHA-512. If that were available, we could extend this API wrapper to support the same or write wrappers for SHA-384/SHA-512 similar to this API.
> 
> Thanks and regards Vivek

Okay and could you pls add test case in commit msg so that we can test it also?

**Comment by VivekUppunda** [comment]
> > > Code LGTM. But why don't we add PSA wrapper to hmac_prf_bits?
> > 
> > 
> > Presently, I am unable to get test vectors for SHA-384 and SHA-512. If that were available, we could extend this API wrapper to support the same or write wrappers for SHA-384/SHA-512 similar to this API.
> > Thanks and regards Vivek
> 
> Okay and could you pls add test case in commit msg so that we can test it also?

We have tested this against WPA2-PSK-256 profiles and the test cases for those using this API have been known to have passed. I have updated the commit message as well

File: src/crypto/crypto_mbedtls_alt.c [hunk 1]
```diff
@@ -605,10 +605,57 @@ int sha384_prf(
...
+{
+    unsigned short ctr, n_le = host_to_le16(buf_len_bits);
+    const u8 *addr[]         = {(u8 *)&ctr, (u8 *)label, data, (u8 *)&n_le};
+    const size_t len[]       = {2, os_strlen(label), data_len, 2};
+    size_t buf_len           = (buf_len_bits + 7) / 8;
+    u8 *buf  = buf_in;
+
+    for (ctr = 1; buf_len >= SHA256_MAC_LEN; buf_len -= SHA256_MAC_LEN, ++ctr)
+    {
+            if (hmac_sha256_vector(key, key_len, 4, addr, len, buf))
```
Comment (fengming-ye) [inline src/crypto/crypto_mbedtls_alt.c:630]: this seems not using addr[i] as input.

File: src/crypto/crypto_mbedtls_alt.c [hunk 2]
```diff
@@ -605,10 +605,57 @@ int sha384_prf(
...
+                    return -1;
+            buf += SHA256_MAC_LEN;
+    }
+
+    if (buf_len)
+    {
+            u8 hash[SHA256_MAC_LEN];
+            if (hmac_sha256_vector(key, key_len, 4, addr, len, hash))
```
Comment (fengming-ye) [inline src/crypto/crypto_mbedtls_alt.c:638]: here also

**Comment by fengming-ye** [comment]
I think the best test is, with the all same input (which can be dumped by debug log), the old version (hmac_prf_bits) and the new version (hmac_sha256_prf) have the same output.
Can you add this only in comments?
Test can refer to crypto_module_tests test_sha256().

File: src/crypto/crypto_mbedtls_alt.c (hunk 1 above)
Comment (VivekUppunda) [reply src/crypto/crypto_mbedtls_alt.c:630]: Kindly refer `sha1_t_prf` or `sha1_prf` in the same file. Also, the internal implementation of sha256 PRF. when the hmac API is called, the number of elements is passed as a parameter to the relevant API. In this case which is 4 elements in the array addr. I have followed the same. The output for the WPA2-PSK-256 connection passed and the test vectors tested also provided desired result

File: src/crypto/crypto_mbedtls_alt.c (hunk 2 above)
Comment (VivekUppunda) [reply src/crypto/crypto_mbedtls_alt.c:638]: Same comment as above comment

**Comment by VivekUppunda** [comment]
> I think the best test is, with the all same input (which can be dumped by debug log), the old version (hmac_prf_bits) and the new version (hmac_sha256_prf) have the same output. Can you add this only in comments? Test can refer to crypto_module_tests test_sha256().

we have collected the test vectors via debug log dump in hmac_prf_bits and verified the same with the new implementation. I will update the commit log as well.

**Review by fengming-ye — APPROVED** [review]

**Review by krish2718 — APPROVED** [review]

**Review by jukkar — APPROVED** [review]

**Pull request merged**